crates/walrus-sdk/src/client/resource.rs

```diff
@@ -519,35 +514,60 @@ impl<'a> ResourceManager<'a> {
                 }
-            } else if let Some(storage_resource) = self
-                .sui_client
-                .owned_storage_for_size_and_epoch(
-                    *encoded_length,
-                    epochs_ahead + self.write_committee_epoch,
-                    &excluded,
-                )
-                .await?
-            {
-                // TODO(joy): Currently select is done one at a time for each blob using `excluded`
-                // to filter, this might not be efficient if the list is too long, consider better
-                // storage selection strategy (WAL-363).
-                // TODO(giac): consider splitting the storage before reusing it (WAL-208).
-                tracing::debug!(
-                    blob_id=%metadata.blob_id(),
-                    storage_object=%storage_resource.id,
-                    "using an existing storage resource to register the blob"
-                );
-                excluded.push(storage_resource.id);
-                reused_metadata_with_storage.push(((*metadata).try_into()?, storage_resource));
-                reused_encoded_lengths.push(*encoded_length);
             } else {
-                tracing::debug!(
-                    blob_id=%metadata.blob_id(),
-                    "no storage resource found for the blob"
-                );
-                new_metadata_list.push(*metadata);
-                new_encoded_lengths.push(*encoded_length);
-            };
+                blob_processing_items.push((*metadata, *encoded_length));
+            }
+        }
+
+        // TODO(joy): Currently select is done one at a time for each blob using `excluded`
+        // to filter, this might not be efficient if the list is too long, consider better
+        // storage selection strategy (WAL-363).
+        // TODO(giac): consider splitting the storage before reusing it (WAL-208).
+        if !blob_processing_items.is_empty() {
+            let all_storage_resources = self
+                .sui_client
+                .owned_storage(ExpirySelectionPolicy::Valid)
+                .await?;
+
+            let target_epoch = epochs_ahead + self.write_committee_epoch;
+            let mut available_resources: Vec<_> = all_storage_resources
+                .into_iter()
+                .filter(|storage| storage.end_epoch >= target_epoch)
+                .collect();
+
+            blob_processing_items.sort_by(|(_, size_a), (_, size_b)| size_b.cmp(size_a));
+            available_resources.sort_by(|a, b| {
+                match a.storage_size.cmp(&b.storage_size) {
+                    std::cmp::Ordering::Equal => a.end_epoch.cmp(&b.end_epoch),
+                    other => other,
+                }
+            });
+
+            for (metadata, encoded_length) in blob_processing_items {
+                let best_resource_idx = available_resources
+                    .iter()
+                    .position(|storage| storage.storage_size >= encoded_length);
```

Comment:
Remark: The current implementation is still `O(m*n)` where `m` is the number of processing items and `n` is the number of resources. I think you should be able to get something like `O(m+n)` by storing the last index into the `available_resources` and moving from there. However, this would be a bit more complicated.

Maybe the easiest would be to skip the sorting of the `available_resources` after all, then you can keep the `swap_remove` below and the code is very simple.

Reply:
applied

Reply:
Suggestion: Change this back to something like the original version:
```suggestion
                let best_resource_idx = available_resources
                    .iter()
                    .enumerate()
                    .filter(|(_, storage)| storage.storage_size >= encoded_length)
                    .min_by(|(_, storage_a), (_, storage_b)| {
                        match storage_a.storage_size.cmp(&storage_b.storage_size) {
                            std::cmp::Ordering::Equal => {
                                storage_a.end_epoch.cmp(&storage_b.end_epoch)
                            }
                            ordering => ordering,
                        }
                    })
                    .map(|(idx, _)| idx);
```

Reply:
I initially used .position() for efficiency (stops early) after sorting, but your explicit .min_by() approach is more robust since it doesn't rely on pre-sorting. I've updated the code with your suggestion.
Thanks for your review